front/context/src/ApiRxContext.tsx

```diff
@@ -9,7 +9,7 @@ import React, { useEffect, useState } from 'react';
 import { ApiRxContextProviderProps } from './types';
 
 export interface ApiRxContextType {
-  api: ApiRx; // From @polkadot/api\
+  api: ApiRx | undefined; // From @polkadot/api\
```

Comment:
Ah, so I wanted to avoid exactly this, in my Zulip comments, since the `api &&` everywhere in the code is kind of useless.

But okay to play along if @yj's okay.

Reply:
I've a hard time building a mental model for that. When you use the `{ api } = useContext(...)`, the `new ApiPromise(..)` (or Rx) is a blocking call and will only return with an `ApiPromise` instance, so that your `api` is never undefined, but you have to wait?

On the other hand, I don't see how we can go around this nicely. if we do:
```js
  const [apiPromise, setApiPromise] = useState<ApiPromise>();
  const [isReady, setIsReady] = useState(false);

  useEffect(() => {
    if (!apiPromise) {
      setApiPromise(new ApiPromise({ provider }));
      setIsReady(false);
    }
  }, [provider]);
```

to prevent the first `useEffect`, I wonder if what will happen when you change provider. This is something you should test but the `apiPromise` might be defined, and there will be no new instantiation.

Reply:
The `new ApiPromise(..)` blocking call is actually not done at the step `{ api } = useContext(...)`, but at the step `<ApiPromiseContextProvider />`, i.e. when you mount the provider component.

Once this component is mounted, and the blocking call is finished, then api will always be defined, so `{ api } = useContext(...)` (instant call, no blocking code) should give a defined api.

Reply:
understood thanks, the `new ApiPromise(..)` is the blocking call, and it's called once as long as the provider isn't changed.